accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+# Inbox Source Generators
+
+**PM** [Immo Landwerth](https://github.com/terrajobst) |
+**Dev** [Eric StJohn](https://github.com/ericstj)
+
+Roslyn source generators allow for component developers to provide functionality that runs in the context of compilation, analyzes a user's code, and can generate additional code to add to the compilation.  Compelling scenarios for source generators involve moving dynamic runtime code to be specialized and statically emitted by the compiler: specializing serializer code for a target type, pre-compiling regular expressions at design time, pre-computing composition graphs for DI systems, and many more.  This technique was possible before but was rather clunky and had to be done outside the compiler, so its use was minimal and not part of the default experience.  Roslyn source generators solve this and as a result we wish to develop many source generators which are part of the default experience for .NET applications.  Source generators present a new type of challenge for a design time component: they need to be part of the toolchain but are tied to specific components within a target framework, and often re-implement or share the functionality of those components.  We need to define how we ship these source generators and make them available to developers.  We need to define the lifecycle for these source generators and how it is tied to the framework component it extends.
```

Comment:
You mention "Roslyn source generators" which include both C# and VB generators yet the runtime is only going to include support for C#. May want to make that a bit clearer.

Reply:
Also what does this mean to F# which is not Roslyn based? Will the user experience when using F# be comparatively worse compared to C#?

Reply:
> yet the runtime is only going to include support for C#

Is this a roslyn decision or one you're imagining we'll make in runtime?  I didn't go into those details here, but I'd expect us to do our best to support all languages possible, similar to analyzers, assuming the compiler supported them.  I have seen samples like @sharwell's that [include VB support](https://github.com/dotnet/roslyn-analyzers/pull/4897/files#diff-fe93da4b6169186079b6aabe3fafe56d70b0f47d0da01e6a825464bfaf1eb20c).

> Will the user experience when using F# be comparatively worse compared to C#

User experience of one language vs others is an interesting one.  I don't expect to cover that here.  This document assumes that the decision has been made to include a generator in the framework.  This could be something discussed for individual generators as they are proposed to the API review board, just like they consider API that is language specific.  FWIW we do spell out that the generator-produced code needs to rely exclusively on public API that meets normal public API compatibility criteria.  This should enable any other component to generate language specific sources to drive any API that supports source-generators.

cc @terrajobst

Reply:
>assuming the compiler supported them

This is one of the limitations impacting the `DllImport` source generator - function pointers. This means VB will not be supported by the interop ones.